# CHANGELOG

We use this CHANGELOG to document breaking changes, new features, bug fixes, and config value changes that may affect both the usage of the workflows and the outputs of the workflows.

## 2025

* 12 June 2025: **Breaking change** phylogenetic - use manually curated `defaults/colors.tsv`
  for lineage colorings. [PR #109][]
* 22 May 2025: Add `benchmark` directives to all workflows. Benchmarks are output
  per rule in workflows in `benchmarks` folder. [PR #108][]
* 19 May 2025: phylogenetic - use multithreading for `augur align` and `augur tree` [PR #105][]
* 31 March 2025: Add host annotations [PR #101][]
    * ingest - new columns `host_genus` and `host_type`
    * phylogenetic - new colorings `host`, `host_genus`, and `host_type`
* 24 March 2025: **Breaking change** ingest - remove config params
  `nextclade.min_length` and `nextclade.min_seed_cover` to just use parameters
  set by the Nextclade dataset [PR #102][]
* 06 March 2025: **Breaking change** phylogenetic and nextclade [PR #99][]
    * move serotype config files to hierarchical directory structure (`defaults/<subtype>/`)
    * move workflow intermediate results to hierarchical directory structure (`results/<subtype>/<gene>/`)
* 27 February 2025: **Breaking change** [PR #93][]
    * ingest
        * switch to community Nextclade dataset "community/v-gen-lab/dengue"
    * phylogenetic
        * split out lineage colorings to `genotype`, `major_lineage`, and `minor_lineage`
* 04 January 2025: **Breaking change** [PR #88][]
    * ingest
        * removed config param `sources`
        * new required config param `curate.genbank_accession`
        * renamed config params:
            * `curate.id_field` to `curate.output_id_field`
            * `curate.sequence_field` to `curate.output_sequence_field`
        * renamed columns:
            * `abbr_authors` to `authors`
            * `authors` to `full_authors`
            * `genbank_accession` to `accession`
    * phylogenetic
        * configs moved from `config` to `defaults`
        * filter out samples with `is_lab_host=true`
        * added frequencies panel configured with the `tip_frequences` params
        * workflow now uses `accession` as id column but defaults to display
          `strain` as the tip label

[PR #88]: https://github.com/nextstrain/dengue/pull/88
[PR #93]: https://github.com/nextstrain/dengue/pull/93
[PR #99]: https://github.com/nextstrain/dengue/pull/99
[PR #101]: https://github.com/nextstrain/dengue/pull/101
[PR #102]: https://github.com/nextstrain/dengue/pull/102
[PR #105]: https://github.com/nextstrain/dengue/pull/105
[PR #108]: https://github.com/nextstrain/dengue/pull/108
[PR #109]: https://github.com/nextstrain/dengue/pull/109

## 2024

* 17 July 2024: Added `.gitattributes` to use LF line endings even on Windows. [PR #81][]
* 16 July 2024: **Breaking change** ingest - replace custom Python scripts
  with `augur curate` commands [PR #78][]
    * New required config param `curate.genbank_location_field`
* 09 July 2024: rename `bin` to `scripts` in all workflows [PR #76][]
* 21 June 2024: nextclade - update example sequences [PR #66][]
* 20 June 2024: nextclade - add outgroup to datasets [PR #67][]
* 05 June 2024: nextclade - fine tuning parameters for datasets [PR #58][]
* 30 May 2024: Add initial draft of nextclade workflow [PR #25][]
* 24 May 2024: **Breaking change** phylogenetic [PR #57][]
    * Use `--root-sequence` for `augur ancestral`
    * new required config param `filter.include`
* 21 May 2024: **Breaking change** Use consistent serotype and genotype names. [PR #51][]
    * ingest
        * column `ncbi_serotype` renamed to `serotype_genbank`
        * column `nextclade_subtype` renamed to `genotype_nextclade`
        * new required config param `curate.serotype_field`
    * phylogenetic
        * in genome/all build, clade membership represents Serotype
        * in genome/denvX build, clade membership represents more detailed Genotype
* 20 May 2024: **Breaking change** phylogenetic - New required config params
  `metadata_url` and `sequences_url` to configure the input URLs. [PR #50][]
* 13 May 2024: phylogenetic - Added E gene builds. See [PR #47][] and [PR #48][]
* 02 May 2024: phylogenetic - restore autogenerated GenBank URLs that was
  missed in previous change. [PR #46][]
* 30 April 2024: **Breaking change** phylogenetic [PR #45][]
    * `augur export` no longer uses `config/auspice_config_{serotype}.json`
    * the Auspice configs are dynamically generated as part of the workflow
    * added `nextclade_subtype` as a default coloring
    * added `assign-colors.py` to generate colors.tsv for `augur export`
* 24 April 2024: ingest - Add gene coverage columns to output metadata. [PR #36][]
* 16 April 2024: phylogenetic - Use inline root sequence. [PR #42][]
* 15 April 2024: phylogenetic [PR #40][]
    * fix input URLs to use latest ingest outputs.
    * added `genbank_accession` to autogenerate GenBank accession URLs in Auspice
* 10 April 2024: ingest - Added annotations to fix serotype from NCBI GenBank. [PR #39][]
* 19 March 2024: **Breaking change** ingest - harmonize with pathogen-repo-guide. [PR #35][]
    * configs are moved from `config` to `defaults`
    * source files moved from `source-data` to `defaults`
    * config param `transform` has been renamed to `curate`
    * Nextstrain automation has been moved to separate build configs
* 26 February 2024: Exclude circular synthetic (or chimeric) and duplicate sequences. [PR #29](https://github.com/nextstrain/dengue/pull/29)
* 23 February 2024: Use nextclade v3 dataset to classify records into subtypes (DENV1/I, DENV1/II, DENV4/I, etc). [PR #16](https://github.com/nextstrain/dengue/pull/16)
* 24 February 2024: Split the pair of metadata.tsv and sequences.fasta files by serotype (DENV1-DENV4) using NCBI `virus-tax-id` field. [PR #20](https://github.com/nextstrain/dengue/pull/20)
* 09 January 2024: Use a phylogenetic directory to start conforming to the [pathogen-repo-guide](https://github.com/nextstrain/pathogen-repo-guide). [PR #15](https://github.com/nextstrain/dengue/pull/15)

[PR #25]: https://github.com/nextstrain/dengue/pull/25
[PR #35]: https://github.com/nextstrain/dengue/pull/35
[PR #36]: https://github.com/nextstrain/dengue/pull/36
[PR #39]: https://github.com/nextstrain/dengue/pull/39
[PR #40]: https://github.com/nextstrain/dengue/pull/40
[PR #42]: https://github.com/nextstrain/dengue/pull/42
[PR #45]: https://github.com/nextstrain/dengue/pull/45
[PR #46]: https://github.com/nextstrain/dengue/pull/46
[PR #47]: https://github.com/nextstrain/dengue/pull/47
[PR #48]: https://github.com/nextstrain/dengue/pull/48
[PR #50]: https://github.com/nextstrain/dengue/pull/50
[PR #51]: https://github.com/nextstrain/dengue/pull/51
[PR #57]: https://github.com/nextstrain/dengue/pull/57
[PR #58]: https://github.com/nextstrain/dengue/pull/58
[PR #66]: https://github.com/nextstrain/dengue/pull/66
[PR #67]: https://github.com/nextstrain/dengue/pull/67
[PR #76]: https://github.com/nextstrain/dengue/pull/76
[PR #78]: https://github.com/nextstrain/dengue/pull/78
[PR #81]: https://github.com/nextstrain/dengue/pull/81

## 2023

* 05 December 2023: Initialize the ingest directory for pulling data from NCBI datasets and subsequent curation for a single pair of metadata.tsv and sequences.fasta files. [PR #13](https://github.com/nextstrain/dengue/pull/13)
* 12 October 2023: Since multiple records can have the same strain name, use the "accession" column as the ID column instead. [PR #12](https://github.com/nextstrain/dengue/pull/12)
* 09 April 2023: Add the DENV2/AII clade amino acid defining changes. [PR #11](https://github.com/nextstrain/dengue/pull/11)

## 2022

* 13 October 2022: Instead of pulling data from the fauna database, pull data from an s3 url [PR #5](https://github.com/nextstrain/dengue/pull/5)
* 05 April 2022: CI: Use a centralized pathogen repo CI workflow. [PR #3](https://github.com/nextstrain/dengue/pull/3)
* 04 April 2022: Migrate CI to GitHub Actions. [PR #3](https://github.com/nextstrain/dengue/pull/3)

## 2021

* 09 November 2021: CI: Upgrade setuptools suite prior to installation. [PR #2](https://github.com/nextstrain/dengue/pull/2)

## 2019

* 10 December 2019: Switch to export v2. [PR #1](https://github.com/nextstrain/dengue/pull/1)
* 16 August 2019: Update build to work with ViPR data. (https://github.com/nextstrain/dengue/commit/081f9f82d971e75848ac1967bf6d841e05428545)

## 2018

* 30 December 2018: Initialize the dengue repository. (https://github.com/nextstrain/dengue/commit/3e0f9feaa4d1799cdda6cf839e03d09390f39c53)
